src/Table.cpp

```diff
@@ -346,4 +346,178 @@ namespace galsim {
 
     template class Table<double,double>;
 
+
+    // ArgGrid
+
+    template<class A>
+    ArgGrid<A>::ArgGrid(const A* args, int N)
+    {
+        grid.reserve(N);
+        for (int i=0; i<N; i++, args++) grid.push_back(*args);
+        const double tolerance = 0.01;
+        da = (grid.back() - grid.front()) / (N-1);
+        if (da == 0.) throw TableError("First and last arguments are equal.");
+        equalSpaced = true;
+        for (int i=1; i<N; i++) {
+            if (std::abs((grid[i] - grid.front())/da - i) > tolerance) equalSpaced = false;
+            if (grid[i] <= grid[i-1])
+                throw TableError("Table2D arguments not strictly increasing.");
+        }
+        lastIndex = 1;
+        lower_slop = (grid[1]-grid[0]) * 1.e-6;
+        upper_slop = (grid[N-1]-grid[N-2]) * 1.e-6;
+    }
+
+    // Look up an index.  Use STL binary search.
+    template<class A>
+    int ArgGrid<A>::upperIndex(const A a) const {
```

Comment:
>A lot of this looks very familiar from the code for Table. For encapsulation purposes, could we modify Table slightly to use ArgGrid as well and avoid the code duplication?

Sure.  I can think of at least 3 paths to potentially take:

  - Get rid of the TableEntry stuff and use an ArgGrid<A> grid and a std::vector<V> vals like in Table2D.  This splits apart the (x, f) pairs though, which complicates some of the inserting/sorting methods.
  - Use an ArgGrid<TableEntry>.  I think this will require exposing a few more of the underlying std::vector methods, (sorting/inserting).
  - Since we're exposing pieces of std::vector, just have ArgGrid publicly inherit from std::vector.

Thoughts?